### DISCUSSION TITLE
Best way to programmatically determine if authenticated

### DISCUSSION THREAD

**Discussion by curelom** [oqtane/oqtane.framework#1654] [Q&A] [answered]
What's best way to determine if authenticated(programmatically?  I've tried using the User object off of the PageState, but the IsAuthenticated proptery on that shows as false when it should be true.

**Comment by hishamco** [comment]
If `User.Identity.IsAuthenticated` is `false` while you are logged in, this is absolutely something wrong or a bug. Please steps to reproduce the issue?

**Reply by curelom** [reply]
I run in WASM mode.  Built a CartComponent inheriting ModuleBase.  I place this component in another ModuleBase component I built (don't know if that's kosher or not and this may be the issue).  On buttonclick in the CarComponent, I then check the pagestate user which shows false even if the user is logged in if(PageState.User.IsAuthenticated) {...}
I don't know if this helps with the bug, but the Roles property on the User object is populated with the proper roles.

**Comment by hishamco** [comment]
Can you try to use `UserService.GetUserAsync()` then check `IsAuthenticated` property for the returned object

**Reply by curelom** [reply]
Thank you for the UserService suggestion.  Unfortunately it also returns false.

**Reply by hishamco** [reply]
Strange!! it's already used in the system, could you show how you use it?

**Comment by sbwalker** [comment, marked as answer]
Oqtane does not rely solely on authentication for security - it also requires users to be members of roles (ie. authorization). The best way to determine if the user is authenticated is to utilize the Oqtane,Security.UserSecurity.IsAuthorized() method. All authenticated users are members of the Registered Users role:

```
UserSecurity.IsAuthorized(PageState.User, RoleNames.Registered)
```

The Page.User.IsAuthenticated property is currently only populated during login to communicate status between the server and client. An enhancement could be added so that this property is also populated by the SiteRouter. This would make the behavior of this property more intuitive and useful. However it would still be recommended to use the UserSecurity.IsAuthorized() method as it provides a higher level of security.

One other item to note... Page.User will be null if the user is not authenticated.

**Reply by sbwalker** [reply]
Page.User.IsAuthenticated is now being populated by the SiteRouter in the Dev branch
